enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,220 @@
+
+## Motivation
+
+Improve Signal to noise of alerts.  Align alert levels "Critical, Warning, Info"
+with operational expectations to allow easier triaging.  Ensure current and
+future alerts have documentation covering problem/investigation/solution steps.
+
+### Goals
+
+* Define clear criteria for designating an alert 'critical'
+* Define clear criteria for designating an alert 'warning'
+* Define minimum time thresholds before prior to triggering an alert
+* Define ownership and responsibilities of existing and new alerts
+* Establish clear documentation guidelines
+* Outline operational triage of firing alerts
+* Explore additional alerting severity (eg, 'minor')
+
+### Non-Goals
+
+* Define specific alert needs
+* Implement user-defined alerts
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an OpenShift developer, I need to ensure that my alerts are appropriately
+tuned to allow end-user operational efficiency.
+
+#### Story 2
+
+As an OpenShift administrator, I need to have confidence that alerts are
+actually informative and actionable to maintain my clusters.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This enhancement will require some developers and administrators to rethink
+their concept of alerts and alert levels.  Many people seem to have an
+intuition about how alerts should work that doesn't necessarily match how
+alerts actually work or how others believe alerts should work.
+
+Formalizing this guidance into a type of template or specification will allow
+developers and administrators to speak a common language and have a common
+understanding, even if the concepts seem initially unituitive.
+
+### Risks and Mitigations
+
+People will make wrong assumptions about alerts and how to handle them.  This
+is not unique to alerts, people often believe certain components should behave
+in a particular way in which they do not.  This is an artifact of poor
+documentation that leaves too much to the imagination.
+
+We can work around this problem with clear and concise documentation.
+
+## Design Details
+
+### Critical Alerts
+
+TL/DR:  For alerting current and impending disaster situations.
+
+Timeline:  ~5 minutes.
+
+Reserve critical level alerts only for reporting conditions that may lead to
+loss of data or inability to deliver service for the cluster as a whole.
+Failures of most individual components should not trigger critical level alerts,
+unless they would result in either of those conditions. Configure critical level
+alerts so they fire before the situation becomes irrecoverable. Expect users to
+be notified of a critical alert within a short period of time after it fires so
+they can respond with corrective action quickly.
+
+Some disaster situations are:
+* loss or impending loss of etcd-quorum
+* etcd corruption
+* inability to route application traffic externally or internally
+(data plane disruption)
+* inability to start/restart any pod other than capacity.  EG, if the API server
```

Comment:
does this imply that the kube-apiserver is considered critical?  Perhaps we should enumerate those binaries which are "special".

Reply:
Anything that can affect an SLA is critical.  kube-apiserver is the only SLA our managed product is contracted to uphold. However, I don't think every alert for kube-apiserver should be set to critical.

Anything that can affect an SLO (which are often company and SRE team specific), should be a warning, because SLO thresholds will be set more aggressively than the SLA that they support.

Reply:
Collapsing discussion from [this earlier thread][1] about this same line, I still don't understand the capacity carve-out.  If a pod failing to schedule is a problem, then the reason the pod failed to schedule helps admins figure out how to resolve/mitigate, but the problem exists regardless of the reason.  E.g. "I have no API-server pods, and am failing to schedule more" is a disaster.  If the reason is "because I'm out of CPU", it's still a disaster.

[1]: https://github.com/openshift/enhancements/pull/637#discussion_r573458919

Reply:
>  E.g. "I have no API-server pods, and am failing to schedule more" is a disaster. If the reason is "because I'm out of CPU", it's still a disaster.

I think this example is contrary to symptom based alerting.  Also, the API-server pods have a very high priority, so capacity should never be an issue on a properly (eg, default or higher) sized node.

Pods failing to schedule due to lack of capacity is not a problem.  Sometimes clusters fill up, sometimes this is desired.  One use case:  run 1 Million of pod x, low priority, preemptible.